subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -691,45 +692,37 @@ private boolean doesConsumerDependOnDestroyer(Node consumer, Node destroyer) {
     }
 
     private static boolean hasOverlap(Iterable<String> paths1, Iterable<String> paths2) {
```

Comment:
The method seems to be used one time only. Have you considered changing the signature to `Set<String>` so it could query the size of the sets and/or use the original sets for exact match lookup?

Reply:
I briefly thought about it but decided to stick with the original signature mostly for conservatism. How is the method usually called? Is one of the sets often empty or much smaller than the other?

Reply:
Frankly speaking, I have no idea. I just thought that you could query `.size()` if you use `Set<...>` arguments.

Reply:
The method is called from two places, and in one of those we only have an Iterable for paths2, not a Set. I looked a bit around the code, and the parameters paths1 can not be empty, this is checked before. Otherwise, I guess the sizes can vary wildly. I added another optimization, skipping the checks when paths2 is empty. Further changes felt like they would complicate the code too much.